src/ray/raylet/reconstruction_policy.h

```diff
@@ -2,33 +2,131 @@
 #define RAY_RAYLET_RECONSTRUCTION_POLICY_H
 
 #include <functional>
+#include <unordered_map>
+#include <unordered_set>
 
+#include <boost/asio.hpp>
+
+#include "ray/gcs/tables.h"
 #include "ray/id.h"
+#include "ray/util/util.h"
+
+#include "ray/object_manager/object_directory.h"
 
 namespace ray {
 
 namespace raylet {
 
-// TODO(swang): Use std::function instead of boost.
+class ReconstructionPolicyInterface {
+ public:
+  virtual void Listen(const ObjectID &object_id) = 0;
+  virtual void Cancel(const ObjectID &object_id) = 0;
+  virtual ~ReconstructionPolicyInterface(){};
+};
 
-class ReconstructionPolicy {
+class ReconstructionPolicy : public ReconstructionPolicyInterface {
  public:
   /// Create the reconstruction policy.
   ///
+  /// \param io_service The event loop to attach reconstruction timers to.
   /// \param reconstruction_handler The handler to call if a task needs to be
   /// re-executed.
-  // TODO(swang): This requires at minimum references to the Raylet's lineage
-  // cache and GCS client.
-  ReconstructionPolicy(std::function<void(const TaskID &)> reconstruction_handler) {}
+  /// \param initial_reconstruction_timeout_ms The initial timeout within which
+  /// a task lease notification must be received. Otherwise, reconstruction
+  /// will be triggered.
+  /// \param client_id The client ID to use when requesting notifications from
+  /// the GCS.
+  /// \param task_lease_pubsub The GCS pub-sub storage system to request task
+  /// lease notifications from.
+  ReconstructionPolicy(boost::asio::io_service &io_service,
+                       std::function<void(const TaskID &)> reconstruction_handler,
+                       int64_t initial_reconstruction_timeout_ms,
+                       const ClientID &client_id,
+                       gcs::PubsubInterface<TaskID> &task_lease_pubsub,
+                       std::shared_ptr<ObjectDirectoryInterface> object_directory);
 
-  /// Check whether an object requires reconstruction. If this object requires
-  /// reconstruction, the registered task reconstruction handler will be called
-  /// for each task that needs to be re-executed.
+  /// Listen for task lease notifications about an object that may require
+  /// reconstruction. If no notifications are received within the initial
+  /// timeout, then the registered task reconstruction handler will be called
+  /// for the task that created the object.
   ///
   /// \param object_id The object to check for reconstruction.
-  void CheckObjectReconstruction(const ObjectID &object_id);
+  void Listen(const ObjectID &object_id);
```

Comment:
I don't think the name `Listen` really captures the meaning of what this method does. It seems more like `ListenAndPossiblyReconstruct`.